### DISCUSSION TITLE
Read `.env` file

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#207] [Q&A] [answered]
What is the recommended way to set environment variables in Marten? It seems that `.env` file is not parsed.

**Comment by ellmetha** [comment, marked as answer]
Hey! Marten does not prescribe a specific method for utilizing environment variables. Developers can choose their preferred approach, whether it's using a `.env` file or any other method, to define environment-specific values. This decision is left to the discretion of developers.

If you want to use `.env` files in your project, I would recommend using the [`cr-dotenv`](https://github.com/gdotdesign/cr-dotenv) shard and simply making sure to initialize settings by reading environment variables using `ENV`. For example:

```crystal
require "dotenv"

Dotenv.load?

Marten.configure do |config|
  config.secret_key = ENV.fetch("MARTEN_SECRET_KEY") { raise "Missing MARTEN_SECRET_KEY env variable" }
end
```

**Reply by notramo** [reply]
Is it documented? This is a pretty basic requirement, it may be worth mentioning in the introduction and the tutorial.

**Reply by ellmetha** [reply]
I don't see the value of documenting this in the introduction or tutorial as these guides mainly exist to help people get started with the framework. Using environment variables to initialize settings is mainly useful when dealing with multiple environments, which usually appears when one needs to deploy the application somewhere. That's why using environment variables to initialize setting values is [already documented in the deployment section](https://martenframework.com/docs/deployment/introduction#secure-critical-setting-values). Using `.env` files specifically is not and I don't consider this to be necessary as this is essentially an extension of using environment variables.

**Reply by notramo** [reply]
There can be many situations where local development requires setting different environment variables between setups, so it can not be hardcoded. For example, multiple developers may configure different Postgres or SMTP testing gateway passwords for their setups.

Also, users may not know about `cr-dotenv`, so they may be clueless where to start.

I would recommend adding a short note like this:

If you want to utilize `.env` for loading environment variables, it's recommended to use the `cr-dotenv` shard.
Example of using it:
```crystal
# config/settings/base.cr
require "dotenv"

Dotenv.load?

# use ENV as usual
config.emailing.from_address = ENV["SMTP_FROM_ADDRESS"]
```

**Reply by ellmetha** [reply]
> There can be many situations where local development requires setting different environment variables between setups, so it can not be hardcoded. For example, multiple developers may configure different Postgres or SMTP testing gateway passwords for their setups.

Definitely, but again: a tutorial aiming at helping people discover the basic features of the framework is not the place to expand on those aspects.

The development section of the documentation would be better suited for that.
